server/src/main/java/org/elasticsearch/plugins/MetadataUpgrader.java

```diff
@@ -15,21 +15,30 @@
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.function.Function;
 import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
 
 /**
  * Upgrades {@link Metadata} on startup on behalf of installed {@link Plugin}s
  */
 public class MetadataUpgrader {
     public final UnaryOperator<Map<String, IndexTemplateMetadata>> indexTemplateMetadataUpgraders;
+    public final Map<String, Function<Metadata.Custom, Metadata.Custom>> customMetadataUpgraders;
 
-    public MetadataUpgrader(Collection<UnaryOperator<Map<String, IndexTemplateMetadata>>> indexTemplateMetadataUpgraders) {
+    public MetadataUpgrader(
+        Collection<UnaryOperator<Map<String, IndexTemplateMetadata>>> indexTemplateMetadataUpgraders,
+        Collection<Map<String, UnaryOperator<Metadata.Custom>>> customMetadataUpgraders
+    ) {
         this.indexTemplateMetadataUpgraders = templates -> {
             Map<String, IndexTemplateMetadata> upgradedTemplates = new HashMap<>(templates);
             for (UnaryOperator<Map<String, IndexTemplateMetadata>> upgrader : indexTemplateMetadataUpgraders) {
                 upgradedTemplates = upgrader.apply(upgradedTemplates);
             }
             return upgradedTemplates;
         };
+        this.customMetadataUpgraders = customMetadataUpgraders.stream()
+            .flatMap(map -> map.entrySet().stream())
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Function::andThen));
```

Comment:
Any reason this uses `andThen` (which will cause a chain of functions), rather than using the iterative approach used for index template upgraders? Both should use the same style - either functions, or iterative.

Reply:
Yeah, fair point. I have changed it to something which iterates over the operators to combine them. It's still not the exactly the same as the templates one, because this one is dealing with a map of operators rather than an operator over maps — this is by design, as here each operator only wants to touch a single custom metadata type. Hopefully this makes sense! I'm including some comments 'cos the collector stuff is a little hairy. I'm also improving the tests in the same commit.